### PR TITLE
Added support for subset copying

### PR DESCRIPTION
**Pull request by prioux**
Includes a new BAC for copying with unregister at the end.
Subset copying works only for two DP types, SSH and S3,
and is only available through the dispatcher API.

Just read the code, point out any problem that is visible by reading it.

### CODE REVIEW

File: BrainPortal/app/models/data_provider.rb [hunk 1]
```diff
@@ -635,6 +635,7 @@ def cache_collection_index(userfile, directory = :all, allowed_types = :regular)
       if userfile.is_a? FileCollection
         if directory == :all
           entries = Dir.glob(userfile.name + "/**/*")
+          #entries = Dir.glob(userfile.name + "/**/*", File::FNM_DOTMATCH).reject { |p| p.ends_with?("/.") || p.ends_with?("/..")}
```
Comment (natacha-beck) [inline BrainPortal/app/models/data_provider.rb:638]: Any reason to have this line commented ?

File: BrainPortal/app/models/s3_flat_data_provider.rb [hunk 2]
```diff
@@ -448,6 +456,28 @@ def rsync_emulation(src_fileinfos,dst_fileinfos) #:nodoc:
...
+    return fileinfos if patterns.blank?
+    patterns = Array(patterns).map(&:to_s)  # e.g. "dir1/dir2/*"
+    fileinfos.select do |fi|
+      finame  = Pathname.new(fi.name)       # "Collection/dir1/dir2/file"
+      fifirst = finame.each_filename.first  # "Collection"
+      next true  if finame.to_s == colname  # "Collection" == "Collection"
+      next false if fifirst     != colname  # should never happen
+      rest = finame.relative_path_from(colname).to_s  # "dir1/dir2/file"
+      selected = patterns.any? { |pat| File.fnmatch(pat, rest, (File::FNM_PATHNAME | File::FNM_DOTMATCH) ) }
+      #if selected
```
Comment (natacha-beck) [inline BrainPortal/app/models/s3_flat_data_provider.rb:472]: Do you want to keep the debug code ?

File: BrainPortal/app/models/ssh_data_provider.rb [hunk 3]
```diff
@@ -202,6 +210,7 @@ def impl_provider_collection_index(userfile, directory = :all, allowed_types = :
       if userfile.is_a? FileCollection
         if directory == :all
           entries = sftp.dir.glob(provider_full_path(userfile).to_s, "**/*")
+          #entries = sftp.dir.glob(provider_full_path(userfile).to_s, "**/*", File::FNM_DOTMATCH)
```
Comment (natacha-beck) [inline BrainPortal/app/models/ssh_data_provider.rb:213]: Should it be in remove ?

**Review by natacha-beck — APPROVED** [review]
Alll, just look to the comment about commented code.

File: BrainPortal/app/models/data_provider.rb (hunk 1 above)
Comment (prioux) [reply BrainPortal/app/models/data_provider.rb:638]: While I was testing all this I discovered a behavior of CBRAIN I might want to change in the future. But I am not sure. And that line is how the behavior would change.

File: BrainPortal/app/models/s3_flat_data_provider.rb (hunk 2 above)
Comment (prioux) [reply BrainPortal/app/models/s3_flat_data_provider.rb:472]: I will remove the debug code once we verified in works correctly.

File: BrainPortal/app/models/ssh_data_provider.rb (hunk 3 above)
Comment (prioux) [reply BrainPortal/app/models/ssh_data_provider.rb:213]: Same thing about the behavior I was saying earlier.

**Pull request merged**